File: R/shiny-alert.R
#' shinyalert
#' 
#' Creates an shinyalert. The show and hide functions are used in server.R. Show
#' an alert placed in ui.R with shinyalert
#' 
#' @param id Specifies the alert id that will be used to access the
#' @param click.hide If TRUE then clicking on the alert will hide it. Defaults to TRUE
#' @param auto.close.after After this many seconds auto close the alert
#'   
#' @family ShinySky elements
#' 
#' @seealso shinyalert
#' 
#' @importFrom shiny tagList singleton tags HTML
#' @examples 
#' 
#' # ui.R
#' 
#' shinyalert("id", click.hide = TRUE, auto.close.after = NULL)
#' 
#' # close after one second
#' # and clicking doesn't hide it
#' shinyalert("id", click.hide = FALSE, auto.close.after = 1)
#' 
#' @export
shinyalert <- function(id, click.hide = TRUE, auto.close.after = NULL) {
    shiny::tagList(shiny::singleton(shiny::tags$head(shiny::tags$script(src = "shinysky/shinyalert.js"))), 
        shiny::HTML(paste0("<div id=\"", id, "\" class=\"shinyalert alert fade\" data-alert=\"alert\" click-hide=\"",as.character(click.hide),"\" data-auto-close-after ='",auto.close.after,"'></div>")))
}

#' showshinyalert
#' 
#' Used in server.R. Show an alert placed in ui.R with shinyalert
#' @param session The Shiny session specified in shinyServer(function(input, output, session) {...}). 
#' @param HTMLtext The HTML you want displayed in the text
#' @param styleclass The styling class of the alert--options are primary, info, 
#'   success ,warning, ,danger ,inverse ,link
#'   
#' @family ShinySky elements
#' @rdname shinyalert
#' @seealso shinyalert
#'   
#' @export
showshinyalert <- function(session, id, HTMLtext, styleclass = "success") {
    alert.css.style = paste("alert", styleclass, sep = "-")
    session$sendCustomMessage("shinyalerthandler", list(id = id, HTMLtext = HTMLtext, alert.css.style = alert.css.style, show=TRUE))
}


#' hideshinyalert
#' 
#' 
#' @family ShinySky elements
#'
#' @rdname shinyalert
#' @seealso shinyalert
#'   
#' @export
hideshinyalert <- function(session, id) {
  session$sendCustomMessage("shinyalerthandler", list(id = id, show=FALSE))
}

File: inst/www/shinyalert.js
 Shiny.addCustomMessageHandler("shinyalerthandler", function(data) {
      var el = $('#' + data.id);
      var classes = el.attr('class').split(' ').filter(function(x) {
           return x.substring(0, 6) === "alert-"
      })

      classes.forEach(function(x) {
           el.removeClass(x)
      })
      
      
      if( data.show ){
        el.addClass(data["alert.css.style"]);
  
        el.addClass('in');
        el.html(data.HTMLtext);
        el.fadeIn();
      } else {
        el.removeClass('in');
        el.addClass('out');
        el.removeClass('fade');
        el.fadeOut();
      }

      //console.log(Number(el.data("auto-close-after")) * 1000); 

      if(el.data("auto-close-after") !== "") {
            setTimeout(function(el) {
              el.removeClass('in');
              el.addClass('out');
              el.removeClass('fade');
              el.fadeOut();
              },Number(el.data("auto-close-after")) * 1000, el)
      }
 });

 // input binding
 $(document).on('click', '.shinyalert', function() {
      var el = $(this);
      if(el.attr("click-hide") === "TRUE") {
            el.removeClass("in")
            el.removeClass('fade')
            el.addClass("out")
            el.fadeOut("slow")            
      }
 });